### ISSUE TITLE
Feedback

### ISSUE THREAD

**Issue by GBH** [Twinpair/VidEdu#1]
Hey. Coming here from Reddit as it's easier to do this here.

You got something going there and that's good. However I can guarantee those tests don't have great coverage. You can check that yourself by running https://github.com/colszowka/simplecov Also attach your project to TravisCi or CircleCi so others can see running test-suite. It's free and it's just a matter of clicking a button.

Small note about style. Personally I hate `test "should whatever"` format. It's really hard to scan the file to see what exactly is being tested. Better would be `test "method name when something"`. When you have a pile of tests for one method it's easy to get an idea what's being tested (or not). I do `def test_some_method`, myself.

Here's absolute minimal tests I'd expect to see that ensure great coverage:

- Model tests:
  - Validation tests. Simple `foo.create(); assert foo.is_invalid?; assert_has_errors_on foo, :param1, param2` goes a long way.
  - Test if you can actually create record. 
  - Test every method you have defined. For example: https://github.com/Twinpair/VidEdu/blob/master/app/models/video.rb has `search`, `get_results`, `order_results` methods are not tested. Those methods have a ton of `if`s there. Make sure you have them covered. Side note. For searching you're probably better off using something like https://github.com/activerecord-hackery/ransack (I see search code in other models. Definitely use that)

- Controller tests:
  - Need to test crud and whatever other actions you got there. On the quick glance your stuff looks good.
  - Rails 5 discourages testing for instance variable assignments but I disagree. Those are data that flows from the controller to the template. I don't give a shit about html structure. I care more about my controller sending the right stuff. There's a plugin that allows for testing that via `assigns` method. Check it out.

- Integration tests:
  - Not used very often. Useful when you want to hit multiple controller/actions in a single session. I use it to test authorization/authentication access for example. I see you have tons of those. They might belong in controller tests. Make sure you don't test the same thing many times. Controller and Integration tests are basically the same thing in Rails 5 now.

About fixtures. Don't do `:one`, `:two`, `:million` in your fixtures. Try to limit yourself to single `:default`. One fixture per model. If you need to test a variation of the default just mutate it with `update_column`. 99% of the tests will only need one fixture. Think of fixtures as a data skeleton of your app. What's the absolute minimum of data you need to have data "populated". For things like searching, ordering, etc it's easy enough to `create` a record you need. Make a helper for that or use a factory like FactoryGirl if you need a lot of that stuff (I prefer my own helpers as I can use fixtures there).

Controllers look way better now.

Gemfile looks good. I'd still do a version lock on all the gems that go to production. 

DB indexes look better. But I'm sure you're missing more. Basically everything you're searching on or sorting needs to be indexed. If you listing videos in chronological order you better have index on `:created_at`.

Remove blank Helper files.

So to make project more presentable I'd
  - Attach it to TravisCI or CircleCI (and add button to the README)
  - Install simpleCov and run it with tests
  - Install Rubocop with Reek and improve code based on what it tells you.

Cheers

**Comment by GBH** [comment]
Also for fixtures you don't need to define `:id`. You can link `has_many` and `belongs` by linking to the fixture name. For example:

```
# users.yml
default:
  name: Tester
```

users have many messages:

```
# messages.yml
default:
  user: default
  content: something
```

**Comment by ericgonzadev** [comment]
Hey @GBH 
Thanks so much for the continuous feedback! The simplecov gem is amazing, I am now covering and hitting 98% of my files with tests. I agree with you, `method name when something` is more readable and provides quick access to what method is being tested. I have updated my test's descriptions to reflect this. I have also limited a singe fixture object per model, with the exception of the user model since I need to test authorization, which requires another user. I have added the necessary indexes to all my db tables. Thanks for introducing me to Rubocop and Reek, they're both neat tools to have when developing apps.
